### DISCUSSION TITLE
Correct way to implement private relay?

### DISCUSSION THREAD

**Discussion by Snarpix** [libp2p/rust-libp2p#5135] [Q&A] [answered]
What is the correct way to implement private relay?

I'm thinking about modifying quic or tls config, to check server and client certificates, but rustls::Server/ClientConfig are private in quic and tls.

I can create a copy of libp2p-tls crate and add client\server cert validation, but it seems non-idiomatic.

**Comment by Snarpix** [comment]
I have found pnet module that does what I want for TCP. 

But how can I use pnet for quic?

```
 let mut swarm = libp2p::SwarmBuilder::with_existing_identity(keypair)
        .with_tokio()
        .with_other_transport(|keypair| {
            let transport = tcp::tokio::Transport::new(tcp::Config::new())
                .and_then(move |socket, _| pnet::PnetConfig::new(psk).handshake(socket));
            let auth = noise::Config::new(keypair)?;
            let mux = yamux::Config::default();

            let tcp_transport = transport
                .upgrade(Version::V1Lazy)
                .authenticate(auth)
                .multiplex(mux);

            Ok(tcp_transport)
        })
        .unwrap()
```

**Comment by gnattu** [comment, marked as answer]
I encountered a similar issue and ended up forking `libp2p-tls` to create my own certification validation mechanism. Upon examining the current implementations, I found that much of the connection handling focuses on ensuring that the other end is `libp2p` based rather than based on `my-program`, which makes it difficult to reject connections from other `libp2p` clients.

A more supported approach would be to implement your own protocol and include it in `libp2p::identify`. Then, you can reject all clients (such as by immediately closing the connection) that do not support your defined protocol. You can use a method to encrypt your protocol string with the PSK, and then let the remote side decrypt this string using the pre-shared key and check if the protocol string matches, if you want to additionally authenticate clients using PSK. However, I'm not a big fan of this approach because I prefer the connection to be rejected by TLS, as tearing down the connections in the `swarm` would trigger many event listeners, which would add unnecessary complexity for me to handle.

Additionally, the concept of `pnet` is somewhat outdated in 2024, and we want to use something better. I don't like the idea of encrypting twice when we can simply reject the remote certificate, but I understand that not all `libp2p` programs are TLS-based, so they need to come up with a solution that works in general. However, if we are implementing something like this we are very likely to have our own program-defined auth and we don't have to care about other programs.

**Comment by thomaseizinger** [comment]
Another option is to use https://github.com/libp2p/rust-libp2p/tree/master/misc/allow-block-list and just define which peers are allowed to connect to you. Or you implement your own connection management by implementing `NetworkBehaviour`.

**Reply by dhuseby** [reply]
This is what I use for my private relays (i.e. a hard coded list of allowed peers since I control the identities all peers use)
